File: ios/LivenessDetection/FaceMetricsExtractor.swift
import Foundation
import CoreGraphics
import MediaPipeTasksVision

enum FaceMetricsExtractor {
  static func extract(result: FaceLandmarkerResult, imageSize: CGSize) -> FaceMetrics? {
    guard let landmarks = result.faceLandmarks.first else { return nil }

    let (leftEar, rightEar) = EyeMouthMetrics.computeEar(landmarks)
    let mar = EyeMouthMetrics.computeMar(landmarks)

    let pose = PoseEstimator.fromTransformationMatrix(
      TransformationMatrixExtractor.extract(from: result)
    ) ?? PoseEstimator.fromLandmarks(landmarks)

    let boundingBox = computeBoundingBox(landmarks, imageSize: imageSize)

    let categories = result.faceBlendshapes.first?.categories ?? []
    func blendScore(_ name: String) -> Float {
      return categories.first(where: { $0.categoryName == name })?.score ?? 0
    }
    let blinkL = blendScore("eyeBlinkLeft")
    let blinkR = blendScore("eyeBlinkRight")
    let blinkScore = (blinkL > 0 || blinkR > 0) ? (blinkL + blinkR) / 2 : 0
    let mouthScore = blendScore("jawOpen")

    var sumX: Float = 0
    var sumY: Float = 0
    for lm in landmarks {
      sumX += lm.x
      sumY += lm.y
    }
    let faceCx = sumX / Float(landmarks.count)
    let faceCy = sumY / Float(landmarks.count)
    let faceSize = distance(landmarks[33], landmarks[263])

    return FaceMetrics(
      yaw: pose.yaw,
      pitch: pose.pitch,
      roll: pose.roll,
      leftEar: leftEar,
      rightEar: rightEar,
      mouthMar: mar,
      blinkScore: blinkScore,
      mouthScore: mouthScore,
      faceCx: faceCx,
      faceCy: faceCy,
      faceSize: faceSize,
      boundingBox: boundingBox,
      timestampMs: Int64(result.timestampInMilliseconds)
    )
  }

  private static func computeBoundingBox(_ landmarks: [NormalizedLandmark], imageSize: CGSize) -> CGRect {
    var minX: Float = 1
    var minY: Float = 1
    var maxX: Float = 0
    var maxY: Float = 0
    for lm in landmarks {
      minX = min(minX, lm.x)
      minY = min(minY, lm.y)
      maxX = max(maxX, lm.x)
      maxY = max(maxY, lm.y)
    }
    return CGRect(
      x: CGFloat(minX) * imageSize.width,
      y: CGFloat(minY) * imageSize.height,
      width: CGFloat(maxX - minX) * imageSize.width,
      height: CGFloat(maxY - minY) * imageSize.height
    )
  }

  private static func distance(_ a: NormalizedLandmark, _ b: NormalizedLandmark) -> Float {
    let dx = a.x - b.x
    let dy = a.y - b.y
    return sqrt(dx * dx + dy * dy)
  }
}

File: android/liveness/src/main/java/com/liveness/detection/FaceMetricsExtractor.kt
package com.liveness.detection

import android.graphics.RectF
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarkerResult
import kotlin.math.max
import kotlin.math.min
import kotlin.math.hypot

object FaceMetricsExtractor {
  fun extract(
    result: FaceLandmarkerResult,
    imageWidth: Int,
    imageHeight: Int,
  ): FaceMetrics? {
    val landmarks = result.faceLandmarks().firstOrNull() ?: return null

    val (leftEar, rightEar) = EyeMouthMetrics.computeEar(landmarks)
    val mar = EyeMouthMetrics.computeMar(landmarks)

    val pose = PoseEstimator.fromTransformationMatrix(
      TransformationMatrixExtractor.extract(result)
    ) ?: PoseEstimator.fromLandmarks(landmarks)

    val boundingBox = computeBoundingBox(landmarks, imageWidth, imageHeight)

    val categories = result.faceBlendshapes().firstOrNull()?.categories() ?: emptyList()
    fun blendScore(name: String): Float {
      return categories.firstOrNull { it.categoryName() == name }?.score() ?: 0f
    }
    val blinkL = blendScore("eyeBlinkLeft")
    val blinkR = blendScore("eyeBlinkRight")
    val blinkScore = if (blinkL > 0f || blinkR > 0f) (blinkL + blinkR) / 2f else 0f
    val mouthScore = blendScore("jawOpen")

    var sumX = 0f
    var sumY = 0f
    for (lm in landmarks) {
      sumX += lm.x()
      sumY += lm.y()
    }
    val faceCx = sumX / landmarks.size
    val faceCy = sumY / landmarks.size
    val faceSize = distance(landmarks[33], landmarks[263])

    return FaceMetrics(
      yaw = pose.yaw,
      pitch = pose.pitch,
      roll = pose.roll,
      leftEyeEar = leftEar,
      rightEyeEar = rightEar,
      mouthMar = mar,
      blinkScore = blinkScore,
      mouthScore = mouthScore,
      faceCx = faceCx,
      faceCy = faceCy,
      faceSize = faceSize,
      boundingBox = boundingBox,
      timestampMs = result.timestampMs(),
      imageWidth = imageWidth,
      imageHeight = imageHeight,
    )
  }

  private fun computeBoundingBox(
    landmarks: List<com.google.mediapipe.tasks.components.containers.NormalizedLandmark>,
    width: Int,
    height: Int,
  ): RectF {
    var minX = 1f
    var minY = 1f
    var maxX = 0f
    var maxY = 0f

    for (lm in landmarks) {
      minX = min(minX, lm.x())
      minY = min(minY, lm.y())
      maxX = max(maxX, lm.x())
      maxY = max(maxY, lm.y())
    }

    return RectF(
      minX * width,
      minY * height,
      maxX * width,
      maxY * height,
    )
  }

  private fun distance(
    a: com.google.mediapipe.tasks.components.containers.NormalizedLandmark,
    b: com.google.mediapipe.tasks.components.containers.NormalizedLandmark,
  ): Float {
    return hypot(a.x() - b.x(), a.y() - b.y())
  }
}
